plugins/multifundchannel.c

```diff
@@ -0,0 +1,1866 @@
+#include <assert.h>
+#include <bitcoin/chainparams.h>
+#include <bitcoin/script.h>
+#include <bitcoin/tx.h>
+#include <ccan/array_size/array_size.h>
+#include <ccan/compiler/compiler.h>
+#include <ccan/crypto/sha256/sha256.h>
+#include <ccan/str/str.h>
+#include <ccan/take/take.h>
+#include <common/addr.h>
+#include <common/amount.h>
+#include <common/features.h>
+#include <common/json.h>
+#include <common/json_stream.h>
+#include <common/jsonrpc_errors.h>
+#include <common/type_to_string.h>
+#include <common/utils.h>
+#include <plugins/libplugin.h>
+#include <stdarg.h>
+#include <string.h>
+
+/* State of whether we have done `fundchannel_start`.  */
+enum multifundchannel_state {
+	/* We have not yet performed `fundchannel_start`.  */
+	MULTIFUNDCHANNEL_START_NOT_YET = 0,
+	/* The `connect` command failed.  `*/
+	MULTIFUNDCHANNEL_CONNECT_FAILED,
+	/* The `fundchannel_start` command succeeded.  */
+	MULTIFUNDCHANNEL_STARTED,
+	/* The `fundchannel_start` command failed.  */
+	MULTIFUNDCHANNEL_START_FAILED,
+	/* The `fundchannel_complete` command failed.  */
+	MULTIFUNDCHANNEL_COMPLETE_FAILED,
+	/* The transaction might now be broadcasted.  */
+	MULTIFUNDCHANNEL_DONE
+};
+
+struct multifundchannel_command;
+
+/* The object for a single destination.  */
+struct multifundchannel_destination {
+	/* The overall multifundchannel command object.  */
+	struct multifundchannel_command *mfc;
+
+	/* The overall multifundchannel_command contains an
+	array of multifundchannel_destinations.
+	This provides the index within the array.
+
+	This is used in debug printing.
+	*/
+	unsigned int index;
+
+	/* ID for this destination.  */
+	const char *id;
+	/* Address hint for this destination, NULL if not
+	specified.
+	*/
+	const char *addrhint;
+	/* The features this destination has.  */
+	const u8 *their_features;
+
+	/* Whether we have `fundchannel_start`, failed `connect` or
+	`fundchannel_complete`, etc.
+	*/
+	enum multifundchannel_state state;
+
+	/* The placeholder address of this destination
+	used during the initial txprepare dryrun.
+	*/
+	const char *placeholder_addr_str;
+	/* The actual target script and address.  */
+	const u8 *funding_script;
+	const char *funding_addr;
+
+	/* The amount to be funded for this destination.
+	If the specified amount is "all" then the `all`
+	flag is set, and the amount is initially 0 until
+	we have figured out how much exactly "all" is,
+	after the dryrun stage.
+	*/
+	bool all;
```

Comment:
By convention `json_to_sat_or_all` uses the privileged value `AMOUNT_SAT(-1ULL)` to mean "all" but I found it a bit wordy to check `dest->amount == AMOUNT_SAT(-1ULL)` instead of just `dest->all`.

Reply:
Also that is `amount_sat`.